documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`Image` uses `RawImage` as-is from Silk.NET.Core, and when set to a non-null value implicitly sets `Style` to custom. As such, you must query `SupportedStyles` before using this property as well. Setting `Image` to `null` will set `Style` back to a standard cursor style, defined by the implementation. It is therefore recommended you set `Style` explicitly when disabling a custom cursor. Note that setting `Style` to a non-`Custom` value will also implicitly set this property to `null`. Setting `Mode` **to** `Custom` explicitly is undefined behaviour, as `Image` won't be set at the time of setting `Mode`.
+
+`SupportedFlags` is a bitmask containing other supported options for the cursor which can be mixed and matched if supported. This must be queried before setting `Flags` - the currently active options. An exception should be thrown if an attempt is made to set an unsupported flag on `Flags`. Unlike the other properties, `Flags` can have multiple bits set.
+
+`Flags` replaces `IsConfined`, which was undefined behaviour on platforms where it was not supported in 2.X - now with `SupportedFlags` you can query its support before setting it.
+
+### Enums
+
+```cs
+[Flags]
+public enum CursorModes
+{
+    Normal,
+    Hidden = 1 << 0,
+    Disabled = 1 << 1,
+    Raw = 1 << 2
+}
+```
+```cs
+[Flags]
+public enum CursorStyles
+{
+    Default,
+    Arrow = 1 << 0,
+    IBeam = 1 << 1,
+    Crosshair = 1 << 2,
+    Hand = 1 << 3,
+    HResize = 1 << 4,
+    VResize = 1 << 5,
+    Custom = 1 << 6,
+}
+```
+```cs
+[Flags]
+public enum CursorFlags
+{
+    None,
+    Confined = 1 << 0
+}
+```
+
+# Keyboard Input
+
+Once again, the interface is very simple.
+
+```cs
+public interface IKeyboard : IInputDevice
+{
+    KeyboardState State { get; }
+    void BeginInput();
+    void EndInput();
+    void SetClipboardText(string? clipboard);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`BeginInput` starts recording textual input, bringing up the on-screen keyboard on platforms where this is needed (i.e. mobile). Unlike 2.0, this is now required given that textual input is now on a full-string basis rather than a per-char basis, and we don't want to record textual input unless the user asks us to.
+
+For instance, instead of a `KeyChar` event being raised every time a character is pressed or repeated, a `TextInput` event is raised as the string is being built alleviating the need for developers to build the string themselves, given that the primary use of this event is for GUI input.
+
+`SetClipboardText` allows setting the text on the user's clipboard so they can paste information from your application in others.
+
+`KeyboardState` is defined as follows:
+
+```cs
+public struct KeyboardState
+{
+    public IKeyboard Device { get; init; }
+    public string? ClipboardText { get; set; } // all sets after the first set forward to Device.SetClipboardText
```

Comment:
I would rather remove the set. It's not obvious what is happening when a property is set but the read value isn't updated.

Reply:
Same response as before - if you think not updating the read value could be confusing then we should make it update the read value.

<sub>
...also I think in any case whether this is a property setter or a set method, users _will_ be confused if ClipboardText doesn't update.</sub>

Reply:
Resolved - forwarding is gone.